### DISCUSSION TITLE
Problem with dither:diffusion-amount=20% with Firebase

### DISCUSSION THREAD

**Discussion by v11** [ImageMagick/ImageMagick#3023] [Help] [answered]
Hello

I am building a small webapp with Firebase to provide e-paper-ready images. To convert images to bitmap i am using a cloud function (node.js).

I have the following string in a Node.js Firebase Cloud Function:
`await spawn('convert', [tempFilePath, '-resize', displaySize + '^', '-gravity', 'center', '-extent', displaySize, dither, ditherValue, '-define', 'dither: diffusion-amount=20%', '-remap', tempFilePath_Map, '-type', 'Palette', 'BMP3:' + tempFilePathBitmap]);
`

Which should be equal to this:
`convert image.pngl -resize 600x448^ -gravity center -extent 600x448 -dither FloydSteinberg -define dither:diffusion-amount=20% -remap eink-7color.png -type Palette, BMP3:images.bmp`

It works very well. But: The following option has no effect:
`dither:diffusion-amount=20%`

Has it something to do with the %-sign? Do i need to escape it correctly? I think Google is is running ImageMagick 7.0.10-48.

Any help is very welcome.

Lukas

**Comment by snibgo** [comment]
> dither: diffusion-amount=20%

Why do you have a space after the colon? Remove it.

For any escapes you might need for Node.js Firebase Cloud, consult its documentation.

**Reply by v11** [reply]
Thank you. But it did not help. I checked the node.js (child-process) documentation and it seems that node.js uses utf8 by default.

Do you know if 
`dither:diffusion-amount=20%`
was available in 7.0.10-48?

Thanks!

**Comment by snibgo** [comment]
"-define dither:diffusion-amount=20%" works fine for me with IM v7.0.8-64 and v7.0.10-45, at the command line.

If you add "-debug command" to the command, IM will send the commands to stderr. For example:
```
2020-12-23T04:16:23+00:00 0:00.424 0.109u 7.0.8 Command magick[5932]: operation.c/CLISettingOptionInfo/440/Command
  at CLI arg 10 - Setting Option: -dither "FloydSteinberg" "(null)"
2020-12-23T04:16:23+00:00 0:00.473 0.109u 7.0.8 Command magick[5932]: operation.c/CLISettingOptionInfo/440/Command
  at CLI arg 12 - Setting Option: -define "dither:diffusion-amount=10%" "(null)"
```
This confirms the percent sign % has been seen. If I deliberately remove the percent sign, the option still has an effect, but the output looks horrible.

**Reply by v11** [reply]
Thanks a lot for testing. I have tried to reproduce it locally and i had no problems to achieve the desired result.
So i assume, i was wrong and Google still runs on Version 6.x of ImageMagick.

Is there a chance to control the diffusion amount of the dither with the existing toolset?

**Comment by snibgo** [comment, marked as answer]
"-define dither:diffusion-amount=10%" has no effect in IM v6.9.9-50.

"magick -version" will write the version information to stdout. I suggest to try that, to establish what version you have.

**Reply by v11** [reply]
I have problem to read from stdout in my Firebase function. I assume, it is version 6. You can close this issue or mark it as answered.

**Comment by snibgo** [comment]
You can close this issue or mark it as answered. I can't.
